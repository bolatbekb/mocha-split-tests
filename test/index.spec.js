"use strict";

const {
  listFiles,
  loadRuntimeStats,
  addKnownRuntimes,
  distributeByRuntime
} = require("../index");

describe("utils", function() {
  describe("listFiles", function() {
    it("should fail when no files are found", function() {
      const pattern = "**/doesnotexist/*.js";
      expect(() => listFiles(pattern)).toThrow(
        `No files matching a pattern: ${pattern}`
      );
    });

    it("should return files when pattern is correct", function() {
      expect(listFiles("*.json")).toEqual(
        expect.arrayContaining(["package.json", "package-lock.json"])
      );
    });
  });

  describe("loadRuntimeStats", function() {
    describe("when runtime.log contains malformed rows", function() {
      it("loads only well-formed rows from file", async function() {
        const stats = await loadRuntimeStats("./test/data/runtimes2.log");
        expect(stats.size).toEqual(3);
        expect(stats).toMatchSnapshot();
      });
    });

    describe("when runtime.log has no malformed rows", function() {
      it("loads all stats from a well-formed file", async function() {
        const stats = await loadRuntimeStats("./test/data/runtimes1.log");
        expect(stats.size).toEqual(11);
        expect(stats).toMatchSnapshot();
      });
    });

    describe("when runtime.log does not exist", function() {
      it("should fail with error", async function() {
        await expect(
          loadRuntimeStats("/invalid/path/to/runtime.log")
        ).rejects.toMatch(
          "no such file or directory, open '/invalid/path/to/runtime.log'"
        );
      });
    });
  });

  describe("addKnownRuntimes", function() {
    it("adds runtime stats to test backlog", function() {
      const testBacklog = [
        "test/1.spec.js",
        "test/2.spec.js",
        "test/3.spec.js"
      ];
      const runtimeStats = new Map()
        .set("test/1.spec.js", 100)
        .set("test/2.spec.js", 200)
        .set("test/4.spec.js", 400);
      expect(addKnownRuntimes(testBacklog, runtimeStats)).toEqual([
        ["test/1.spec.js", 100],
        ["test/2.spec.js", 200],
        ["test/3.spec.js", 0]
      ]);
    });
  });

  describe("distributeByRuntime", function() {
    it("evenly distributes test files between buckets", async function() {
      const runtimeStats = new Map()
        .set("1", 100)
        .set("2", 200)
        .set("3", 300)
        .set("4", 400)
        .set("5", 500)
        .set("6", 600)
        .set("7", 700)
        .set("8", 800)
        .set("9", 900);
      const testFiles = ["1", "2", "3", "4", "5", "6", "7", "8", "9"];
      const result = await distributeByRuntime({
        testFiles,
        runtimeStats,
        totalGroups: 4
      });
      expect(result).toMatchSnapshot();
    });
  });
});
