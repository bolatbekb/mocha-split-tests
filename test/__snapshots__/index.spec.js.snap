// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils distributeByRuntime evenly distributes test files between buckets 1`] = `
Array [
  Object {
    "files": Array [
      "9",
      "2",
      "1",
    ],
    "runtime": 1200,
  },
  Object {
    "files": Array [
      "8",
      "3",
    ],
    "runtime": 1100,
  },
  Object {
    "files": Array [
      "7",
      "4",
    ],
    "runtime": 1100,
  },
  Object {
    "files": Array [
      "6",
      "5",
    ],
    "runtime": 1100,
  },
]
`;

exports[`utils loadRuntimeStats when runtime.log contains malformed rows loads only well-formed rows from file 1`] = `
Map {
  "test/controllers/application_controller.spec.js" => 5084,
  "test/models/user.spec.js" => 5084,
  "test/does-not-exist-anymore.spec.js" => 234,
}
`;

exports[`utils loadRuntimeStats when runtime.log has no malformed rows loads all stats from a well-formed file 1`] = `
Map {
  "test/controllers/application_controller.spec.js" => 5084,
  "test/models/user.spec.js" => 5084,
  "test/models/company.spec.js" => 1234,
  "test/routers/foo.spec.js" => 20,
  "test/config/config.spec.js" => 345,
  "test/models/account.spec.js" => 2435,
  "test/routers/bar.spec.js" => 20,
  "test/config/buz.spec.js" => 345,
  "test/models/aaa.spec.js" => 2435,
  "test/does-not-exist-anymore.spec.js" => 234,
  "test/does-not-exist-anymore_2.spec.js" => 234234,
}
`;
